clusterloader2/pkg/measurement/common/network/network_performance_measurement.go

```diff
@@ -0,0 +1,474 @@
+	dynamicClient := npm.framework.GetDynamicClients().GetClient()
+	gvk := schema.GroupVersionKind{Group: customResourceGroup, Kind: customResourceKind, Version: customResourceVersion}
+	gvr, _ := meta.UnsafeGuessKindToResource(gvk)
+	watchResource, err := dynamicClient.Resource(gvr).Watch(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		klog.Error("error watching", err)
+		return
+	}
+	eventChannel := watchResource.ResultChan()
+	for event := range eventChannel {
+		var metricResp metricResponse
+		if event.Type != watch.Modified {
+			continue
+		}
+		unstructuredList, _ := event.Object.(runtime.Unstructured)
+		unstructuredObjContent := unstructuredList.UnstructuredContent()
+		metricMap := unstructuredObjContent["status"].(map[string]interface{})
+		if err = getMetricResponse(metricMap, &metricResp); err != nil {
+			klog.Error(err)
+			continue
+		}
+		if len(metricResp.Metrics) == 0 && metricResp.Error != "" {
+			klog.Error("error message from worker pod: ", metricResp.Error)
+			continue
+		}
+		npm.metricLock.Lock()
+		npm.metricVal = append(npm.metricVal, metricResp.Metrics)
+		npm.metricLock.Unlock()
+	}
+}
+
+func (npm *networkPerformanceMeasurement) execNToMTest() {
+	go npm.createCustomResourcePerUniquePodPair()
+
+	npm.formUniquePodPair()
+}
+
+func (npm *networkPerformanceMeasurement) createCustomResourcePerUniquePodPair() {
+	for CRIndex := 0; ; CRIndex++ {
+		uniqPodPair := <-npm.podPairCh
+		if uniqPodPair.isLastPodPair {
+			break
+		}
+		if npm.startTimeStampForTestExecution == 0 {
+			npm.startTimeStampForTestExecution = getStartTimeStampForPod()
+		}
+		npm.templateMapping = npm.populateTemplate(uniqPodPair, CRIndex)
+		if err := npm.framework.ApplyTemplatedManifests(networkTestRequestFilePath, npm.templateMapping); err != nil {
+			klog.Error(err)
+		}
+		npm.podPairNumber++
+	}
+}
+
+func (npm *networkPerformanceMeasurement) formUniquePodPair() {
+	var uniqPodPair uniquePodPair
+	lastPodPair := uniquePodPair{isLastPodPair: true}
+
+	var unUsedPodIndex = 0
+	for {
+		for workerNode, podList := range npm.workerPodList {
+			unUsedPod, err := getUnusedPod(&podList)
+			npm.workerPodList[workerNode] = podList
+			if err != nil {
+				delete(npm.workerPodList, workerNode)
+				continue
+			}
+			unUsedPodIndex++
+
+			if unUsedPodIndex == 1 {
+				uniqPodPair.sourcePodIP = unUsedPod.podIP
+				uniqPodPair.sourcePodName = unUsedPod.podName
+			} else if unUsedPodIndex == 2 {
+				uniqPodPair.destinationPodIP = unUsedPod.podIP
+				uniqPodPair.destinationPodName = unUsedPod.podName
+				unUsedPodIndex = 0
+				npm.podPairCh <- uniqPodPair
+			}
+		}
+		if len(npm.workerPodList) == 0 {
+			npm.podPairCh <- lastPodPair
+			break
+		}
+	}
+}
+
+func (npm *networkPerformanceMeasurement) populateWorkerPodList(data *workerPodData) {
+	if podData, ok := npm.workerPodList[data.workerNode]; !ok {
+		npm.workerPodList[data.workerNode] = []workerPodData{{podName: data.podName, workerNode: data.workerNode, podIP: data.podIP}}
```

Comment:
the convention is generally:

if _, ok := ...; !ok {
  npm.workerPodList[...] = []workerPodData{}
}
npm.workerPodList[...] = append(npm.workerPodList[...], *data)

Reply:
Modified as per suggestion.